File: sources/opencv-samples/2_ReadAndWriteImage/main.py
# https://docs.opencv.org/4.x/db/deb/tutorial_display_image.html
# https://github.com/opencv/opencv/blob/4.x/samples/python/tutorial_code/introduction/display_image/display_image.py

import cv2 as cv
import sys
import argparse


def main():

    args = parser.parse_args()

    path = cv.samples.findFile(args.image_path)
    if not path:
        sys.exit(f"Image {args.image_path} not found")

    if not args.saved_image_path:
        sys.exit(f"Path {args.saved_image_path} is not valid")

    mode = cv.IMREAD_UNCHANGED

    if args.mode:
        mode_name = ""
        if args.mode == "u" or args.mode == "U":
            mode_name = "IMREAD_UNCHANGED"
        elif args.mode == "c" or args.mode == "C":
            mode_name = "IMREAD_COLOR"
            mode = cv.IMREAD_COLOR
        elif args.mode == "g" or args.mode == "G":
            mode_name = "IMREAD_GRAYSCALE"
            mode = cv.IMREAD_GRAYSCALE
        else:
            print("Unexpected mode value")
            print("Read the README.md")
            sys.exit()
        print(f"Read mode: {mode_name}")

    img = cv.imread(path, mode)

    if img is None:
        sys.exit("Could not read the image")

    cv.imshow("Read Image", img)
    cv.imwrite(args.saved_image_path, img)
    print(f"Image saved to {args.saved_image_path}")
    print("Press a key to quit")
    cv.waitKey(0)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="Read and Write Image main.py CLI")
    parser.add_argument("image_path", type=str, help="path to image file")
    parser.add_argument("saved_image_path", type=str, help="path to image file to save")
    parser.add_argument("-m", "--mode", type=str, help="mode to read the image")
    main()

File: sources/opencv-samples/2_ReadAndWriteImage/main.cpp
// https://docs.opencv.org/4.x/db/deb/tutorial_display_image.html
// https://github.com/opencv/opencv/blob/4.x/samples/cpp/tutorial_code/introduction/display_image/display_image.cpp

#include <opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
using namespace cv;
using namespace std;

int main(int argc, char** argv)
{
    string image_path;
    string saved_image_path;
    string mode_name;
    Mat img;

    if (argc < 3 || argc > 4) 
    {
        printf("Usage: ./main.out <image_path> <saved_image_path> [mode]\n");
        printf("Read the README.md\n");
        return -1;
    }

    image_path = samples::findFile(argv[1]);
    saved_image_path = argv[2];

    cv::ImreadModes mode = IMREAD_UNCHANGED;

    if (argc == 4)
    {   
        if (!strcmp(argv[3], "u") || !strcmp(argv[3], "U"))
        { 
            mode_name = "IMREAD_UNCHANGED";
        }
        else if (!strcmp(argv[3], "c") || !strcmp(argv[3], "C"))
        {
            mode =  IMREAD_COLOR;
            mode_name = "IMREAD_COLOR";
        }
        else if (!strcmp(argv[3], "g") || !strcmp(argv[3], "G"))
        {
            mode = IMREAD_GRAYSCALE;
            mode_name = "IMREAD_GRAYSCALE";
        }
        else
        {
            printf("Unexpected mode value\n");
            printf("Read the README.md\n");
            return -1;
        }
        cout << "Read mode: " << mode_name << endl;
    }

    img = imread(image_path, mode);

    if(img.empty())
    {
        cout << "Could not read the image: " << image_path << endl;
        return -1;
    }
    imshow("Read Image", img);
    /*
    printf("Press 's' to save the image or any other key to quit.\n");
    int k = waitKey(0); // Wait for a keystroke in the window
    if(k == 's')
    {
        printf("Image saved to ./img1.png\n");
        imwrite("img1.png", img);
    }*/
    cout << "Image saved to " << saved_image_path << endl;
    imwrite(saved_image_path, img);
    cout << "Press a key to quit" << endl;
    waitKey(0);
    cv::destroyAllWindows();
    return 0;
}